llvm/lib/Target/X86/X86InstrInfo.cpp

```diff
@@ -10333,5 +10333,15 @@ void X86InstrInfo::genAlternativeCodeSequence(
   }
 }
 
+// See also: X86DAGToDAGISel::SelectInlineAsmMemoryOperand().
+void X86InstrInfo::getFrameIndexOperands(SmallVectorImpl<MachineOperand> &Ops) const {
+  Ops.append({
+    MachineOperand::CreateImm(1),        // Scale
+    MachineOperand::CreateReg(0, false), // Index
+    MachineOperand::CreateImm(0),        // Disp
+    MachineOperand::CreateReg(0, false), // Segment
+  });
```

Comment:
I was wondering if this could/should be expressed with `X86AddressMode`.
`X86AddressMode M; M.getFullAddress(Ops)` seems to be very close to what this is doing. Except that this API here does not explicitely add a "base" (reg or frame index), but maybe it should start doing that? (So in `foldInlineAsmMemOperand` just remove the register operand (instead of using `ChangeToFrameIndex`) and have this function add the frame index operand as well. That would also help if a target has a different order of operands...

Reply:
oh! yeah it does look potentially reusable.

Though the class seems very x86 specific.  Would be nice if there was a target independent interface I could call from `foldInlineAsmMemOperand` (added in) https://github.com/llvm/llvm-project/pull/70743/files to get this info.

Let me check the users of `X86AddressMode`... worst case I could call it and drop the first operand.

> and have this function add the frame index operand as well. That would also help if a target has a different order of operands...

I guess if `X86AddressMode` isn't a per-target interface, then in `foldInlineAsmMemOperand` I'd have to do some target specific hack, then we back to adding this interface?

Reply:
> I guess if X86AddressMode isn't a per-target interface, then in foldInlineAsmMemOperand I'd have to do some target specific hack

example of what I roughly have in mind
```diff
diff --git a/llvm/lib/CodeGen/TargetInstrInfo.cpp b/llvm/lib/CodeGen/TargetInstrInfo.cpp
index 5ede36505b5b..bfbe477f97a3 100644
--- a/llvm/lib/CodeGen/TargetInstrInfo.cpp
+++ b/llvm/lib/CodeGen/TargetInstrInfo.cpp
@@ -578,13 +578,21 @@ static void foldInlineAsmMemOperand(MachineInstr *MI, unsigned OpNo, int FI,
     foldInlineAsmMemOperand(MI, TiedTo, FI, TII);
   }
 
+  SmallVector<MachineOperand, 4> NewOps;
+  MachineOperand *InsertPt;
+  if (isa<X86GenInstrInfo>(TII)) {
+    X86AddressMode M;
+    M.getFullAddress(NewOps);
+    InsertPt = MI->operands_begin() + OpNo;
+  } else {
   // Change the operand from a register to a frame index.
-  MO.ChangeToFrameIndex(FI, MO.getTargetFlags());
+    MO.ChangeToFrameIndex(FI, MO.getTargetFlags());
+    TII.getFrameIndexOperands(NewOps);
+    InsertPt = MI->operands_begin() + OpNo + 1;
+  }
 
-  SmallVector<MachineOperand, 4> NewOps;
-  TII.getFrameIndexOperands(NewOps);
   assert(!NewOps.empty() && "getFrameIndexOperands didn't create any operands");
-  MI->insert(MI->operands_begin() + OpNo + 1, NewOps);
+  MI->insert(InsertPt, NewOps);
 
   // Change the previous operand to a MemKind InlineAsm::Flag. The second param
   // is the per-target number of operands that represent the memory operand

```

Reply:
Yes addressing modes are highly target specific. I was just proposing it's use for the implementation within `X86InstrInfo::getFrameIndexOperands`

Reply:
Sorry for the excessive delays following up on this; have been traveling for conference season for the past two weeks.  I've just not made the changes you've suggested, and I think I do indeed like it better.  Will post a commit on top once my tests finish successfully locally.

Reply:
done in https://github.com/llvm/llvm-project/pull/70832/commits/06c300b3a95d679107e97191a535672ada75190c PTAL